resqpy/olio/relperm.py

```diff
@@ -0,0 +1,316 @@
+
+   def write_hdf5_and_create_xml(self):
+      """Write relative permeability table data to hdf5 file and create xml for RESQML
+        objects to represent dataframe."""
+
+      super().write_hdf5_and_create_xml()
+      # note: time series xml must be created before calling this method
+      mesh_root = self.mesh.root
+      # create an xml of extra metadata to indicate that this is a relative permeability table
+      rqet.create_metadata_xml(mesh_root, {'relperm_table': 'true'})
+
+
+def text_to_relperm_dict(filepath):
+   """
+    Returns a dictionary that contains dataframes with relative permeability and capillary pressure data and
+    phase combinations.
+
+    arguments:
+    filepath (str): relative or full path of the text file to be processed
+
+    returns:
+    dict, each element in the dictionary contains a dataframe, with saturation and rel. permeability/capillary pressure
+    data, and the phase combination being described
+    """
+   with open(filepath) as f:
+      # create list of rows of the original ascii file with blank lines removed
+      data = list(filter(None, [list(filter(None, x.strip('\n').split(' '))) for x in f.readlines()]))
+      # remove comments
+      data = [x for x in data if ('!' not in x)]
+
+   # get indices of start of each new relperm table based on Nexus keywords
+   table_start_positions = [
+      i for i, l in enumerate(data) if len({'WOTABLE', 'GOTABLE', 'GWTABLE'}.intersection(set(l))) == 1
+   ]
+   df_cols_dict = {
+      'SW': 'Sw',
+      'SG': 'Sg',
+      'KRW': 'Krw',
+      'KRG': 'Krg',
+      'KROW': 'Kro',
+      'KROG': 'Kro',
+      'KRWG': 'Krw',
+      'PCWO': 'Pc',
+      'PCGO': 'Pc',
+      'PCGW': 'Pc'
+   }
+   relperm_table_idx = 1
+   rel_perm_dict = {}
+   for i, l in enumerate(table_start_positions):
+      key = 'relperm_table' + str(relperm_table_idx)
+      rel_perm_dict[key] = {}
+      relperm_table_idx += 1
+      if 'WOTABLE' in data[l]:
+         phase_combo = 'water-oil'
+         rel_perm_dict[key]['phase_combo'] = phase_combo
+      elif 'GOTABLE' in data[l]:
+         phase_combo = 'gas-oil'
+         rel_perm_dict[key]['phase_combo'] = phase_combo
+      elif 'GWTABLE' in data[l]:
+         phase_combo = 'gas-water'
+         rel_perm_dict[key]['phase_combo'] = phase_combo
+      else:
+         raise Exception('incorrect table key word encountered')
+      if i < (len(table_start_positions) - 1):
+         table_end = table_start_positions[i + 1]
+      else:
+         table_end = len(data)
+      table_cols = data[l + 1]
+      table_rows = data[l + 2:table_end]
+      df = pd.DataFrame(table_rows, columns = table_cols)
+      df.columns = df.columns.map(df_cols_dict)
+      sat_col = [x for x in df.columns if 'S' in x][0]
+      df = df.apply(pd.to_numeric, errors = 'coerce')
+      df = df[df[sat_col].notnull()]
+      rel_perm_dict[key]['df'] = df
+   return rel_perm_dict
+
+
+def relperm_parts_in_model(model, title = None, related_uuid = None):
```

Comment:
If the 3 extra items are added as extra metadata, they could also be optional arguments to this method, so calling code could say (for example) _low_sal = True_ to limit the returned parts to low salinity tables. The default of None would need to cause that item to be left out of the dictionary passed to model.parts()

Reply:
The method now includes the option to narrow the search based on these 3 extra metadata items :)